### PR TITLE
Fix RPS LJE and update documentation

### PR DESCRIPTION
**Pull request by Agustin-Picard**
# Fix RPS LJE and update documentation

Although it provided satisfactory results in the CIFAR10 benchmark, the implementation of RPS LJE didn't actually follow the description in the [paper](https://proceedings.neurips.cc/paper/2021/file/c460dc0f18fc309ac07306a4a55d2fd6-Paper.pdf). This is fixed in 55bffb191a10847943d43133c16bb93c1c778f99 . An abstraction is also implemented to simplify the shared code between RPS L2 and RPS LJE in c28abb6750443191f4d8f96b7877602da99fae77 . 

The tests were adjusted to follow the current implementation.

I was forced to opt for a less optimized version in 727e5e4f0f406b499b399a5cbbeb9392d292513e without end-to-end lazy computations for the estimation of the hessian as it led to cuBLAS errors when applying the pinv operation.

#21 is also fixed by removing the abstract method without the actual implementation in dev@f1a360f

More thorough documentation was added throughout and more notebooks are included in the README.

### CODE REVIEW

File: deel/influenciae/benchmark/influence_factory.py [hunk 1]
```diff
@@ -202,9 +202,7 @@ def build(self, training_dataset: tf.data.Dataset, model: tf.keras.Model,
             dataset_hessian = training_dataset
         else:
             batch_size = training_dataset._batch_size.numpy()  # pylint: disable=W0212
-            take_size = int(
-                np.ceil(float(self.dataset_hessian_size) / batch_size)) * batch_size
-            dataset_hessian = training_dataset.take(take_size)
+            dataset_hessian = training_dataset.unbatch().take(self.dataset_hessian_size).batch(batch_size)
```
Comment (lucashervier) [inline deel/influenciae/benchmark/influence_factory.py:205]: If we have to unbatch it, isn't it better to ask the user to provide a dataset that is not batched in the first place?

File: deel/influenciae/boundary_based/sample_boundary.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ def __delta_to_index(indexes_1: tf.Tensor, indexes_2: tf.Tensor, x: tf.Tensor):
         return delta_x
 
     @tf.function
-    def __step(self, x: tf.Tensor, y_pred: tf.Tensor) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
+    def _step(self, x: tf.Tensor, y_pred: tf.Tensor) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
```
Comment (lucashervier) [inline deel/influenciae/boundary_based/sample_boundary.py:78]: Why making it protected instead of private? :)

File: deel/influenciae/common/base_influence.py [hunk 3]
```diff
@@ -269,30 +269,6 @@ def compute_influence_vector(
 
         return inf_vect_ds
 
-    @abstractmethod
```
Comment (lucashervier) [inline deel/influenciae/common/base_influence.py:272]: Why removing it? Not all class should implement it?

File: deel/influenciae/rps/rps_l2.py [hunk 4]
```diff
@@ -59,158 +59,17 @@ def __init__(
             lambda_regularization: float,
             scaling_factor: float = 0.1,
             epochs: int = 100,
-            layer_index: int = -2,
+            layer_index: int = -1,
```
Comment (lucashervier) [inline deel/influenciae/rps/rps_l2.py:62]: Considering the theorem, isn't it better that this value is always -1? (*i.e.* not an init param)

File: deel/influenciae/rps/rps_lje.py [hunk 5]
```diff
@@ -52,81 +47,112 @@ def __init__(
...
-
-        # Compute weight factor for the optimization step
-        size = tf.data.experimental.cardinality(dataset)
-        iter_dataset = iter(influence_vector_dataset)
-        weight_size = tf.reduce_sum(
-            tf.stack([tf.reduce_prod(tf.shape(w)) for w in influence_model.model.layers[target_layer].weights])
+        super().__init__(influence_model.model, dataset, influence_model.loss_function)
+        self.epsilon = tf.constant(epsilon, dtype=tf.float32)
+
+        # In the paper, the authors explain that in practice, they use a single step of SGD to compute the
```
Comment (lucashervier) [inline deel/influenciae/rps/rps_lje.py:56]: Later, you say you accumulate the gradient, so i guess it is a single GD step no?

**Review by lucashervier — COMMENTED** [review]
LGTM! I have a few questions as you will see in the comments. Nothing that requires change but where I am wondering about certain choices

File: deel/influenciae/benchmark/influence_factory.py (hunk 1 above)
Comment (Agustin-Picard) [reply deel/influenciae/benchmark/influence_factory.py:205]: We could do that, but the problem is that these factories are being used inside `InfluenceCalculator`s, where the datasets are actually asked to be batched.
The change here is that when applying the `take` method, it is no longer recognized as a batched dataset and doesn't pass some asserts further in the code (inside RPS LJE)

File: deel/influenciae/common/base_influence.py (hunk 3 above)
Comment (Agustin-Picard) [reply deel/influenciae/common/base_influence.py:272]: Because it was already implemented further down. I was just addressing issue #21 ;)

File: deel/influenciae/rps/rps_l2.py (hunk 4 above)
Comment (Agustin-Picard) [reply deel/influenciae/rps/rps_l2.py:62]: Because of the way I'm splitting the model into two graphs in the new abstraction's constructor, it gives the same results on the tests that I've run.

File: deel/influenciae/rps/rps_lje.py (hunk 5 above)
Comment (Agustin-Picard) [reply deel/influenciae/rps/rps_lje.py:56]: Yup, it's a single gradient step on the amount of samples specified by `n_samples_for_hessian`.

Comment (lucashervier) [reply deel/influenciae/rps/rps_lje.py:56]: Then, should not we change the comment or do you think it's overkill?

File: deel/influenciae/benchmark/influence_factory.py (hunk 1 above)
Comment (lucashervier) [reply deel/influenciae/benchmark/influence_factory.py:205]: Ok, I guess it is alright for now but we should be careful of user feedback if it is causing trouble!

**Review by lucashervier — APPROVED** [review]
All conversation are resolved now. Congrats!
